src/Processors/Transforms/WindowTransform.cpp

```diff
@@ -156,34 +484,158 @@ void WindowTransform::transform(Chunk & chunk)
 
-            if (new_partition)
+            a->create(buf);
+        }
+    }
+}
+
+IProcessor::Status WindowTransform::prepare()
+{
+//    fmt::print(stderr, "prepare, next output {}, not ready row {}, first block {}, hold {} blocks\n",
+//        next_output_block_number, first_not_ready_row, first_block_number,
+//        blocks.size());
+
+    if (output.isFinished())
+    {
+        // The consumer asked us not to continue (or we decided it ourselves),
+        // so we abort.
+        input.close();
+        return Status::Finished;
+    }
+
+    assert(first_not_ready_row.block >= first_block_number);
+    // The first_not_ready_row might be past-the-end if we have already
+    // calculated the window functions for all input rows. That's why the
+    // equality is also valid here.
+    assert(first_not_ready_row.block <= first_block_number + blocks.size());
+    assert(next_output_block_number >= first_block_number);
+
+    // Output the ready data prepared by work().
+    // We inspect the calculation state and create the output chunk right here,
+    // because this is pretty lightweight.
+    if (next_output_block_number < first_not_ready_row.block)
+    {
+        if (output.canPush())
+        {
+            // Output the ready block.
+//            fmt::print(stderr, "output block {}\n", next_output_block_number);
+            const auto i = next_output_block_number - first_block_number;
+            ++next_output_block_number;
+            auto & block = blocks[i];
+            auto columns = block.input_columns;
+            for (auto & res : block.output_columns)
             {
-                // Reset the aggregate function states.
-                a->destroy(buf);
-                a->create(buf);
+                columns.push_back(ColumnPtr(std::move(res)));
             }
+            output_data.chunk.setColumns(columns, block.numRows());
+
+            output.pushData(std::move(output_data));
+        }
 
-            // Update the aggregate function state and save the result.
-            a->add(buf,
-                ws.argument_columns.data(),
-                row,
-                arena.get());
+        // We don't need input.setNotNeeded() here, because we already pull with
+        // the set_not_needed flag.
+        return Status::PortFull;
+    }
 
-            a->insertResultInto(buf,
-                *ws.result_column,
-                arena.get());
+    if (input_is_finished)
+    {
+        // The input data ended at the previous prepare() + work() cycle,
+        // and we don't have ready output data (checked above). We must be
+        // finished.
+        assert(next_output_block_number == first_block_number + blocks.size());
+        assert(first_not_ready_row == blocksEnd());
+
+        // FIXME do we really have to do this?
+        output.finish();
+
+        return Status::Finished;
+    }
+
+    // Consume input data if we have any ready.
+    if (!has_input && input.hasData())
```

Comment:
Is the variable `has_input` required? I see the `input.pullData(true)` has set the input as 'not needed' and the status is Ready. Is it any possibility that another input data block comes before `work()` executes? There is a queue `blocks` for input data blocks in `work()`.